### PR TITLE
Extended contexts to regression and multiclass, added FFM pigstension

### PR DESCRIPTION
**Pull request by Zruty0**
Part of #754, extends #949.
Added MulticlassClassification context and RegressionContext, with corresponding Evaluate methods. Also added FFM binary trainer to the context extensions.

### CODE REVIEW

File: src/Microsoft.ML.Data/Evaluators/MulticlassClassifierEvaluator.cs [hunk 1]
```diff
@@ -598,62 +598,38 @@ internal Result(IExceptionContext ectx, IRow overallResult, int topK)
...
-        /// <param name="pred">The index delegate for columns from prediction of a multi-class classifier.
-        /// Under typical scenarios, this will just be the same tuple of results returned from the trainer.</param>
-        /// <param name="topK">If given a positive value, the <see cref="Result.TopKAccuracy"/> will be filled with
-        /// the top-K accuracy, that is, the accuracy assuming we consider an example with the correct class within
-        /// the top-K values as being stored "correctly."</param>
+        /// <param name="data">The scored data.</param>
+        /// <param name="label">The name of the label column in <paramref name="data"/>.</param>
+        /// <param name="score">The name of the score column in <paramref name="data"/>.</param>
+        /// <param name="predictedLabel">The name of the predicted label column in <paramref name="data"/>.</param>
+        /// <returns>The evaluation results for these uncalibrated outputs.</returns>
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/Evaluators/MulticlassClassifierEvaluator.cs]: >returns [](start = 13, length = 7)

remove

File: test/Microsoft.ML.StaticPipelineTesting/Training.cs [hunk 2]
```diff
@@ -2,15 +2,17 @@
...
 using Microsoft.ML.Runtime;
 using Microsoft.ML.Runtime.Data;
 using Microsoft.ML.Runtime.Internal.Calibration;
 using Microsoft.ML.Runtime.Learners;
+using Microsoft.ML.Runtime.RunTests;
 using Microsoft.ML.Runtime.Training;
 using System;
 using Xunit;
 using Xunit.Abstractions;
+using Microsoft.ML.Runtime.FactorizationMachine;
```
Comment (Zruty0) [inline test/Microsoft.ML.StaticPipelineTesting/Training.cs]: >using [](start = 0, length = 5)

(placeholder)

File: src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineStatic.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Training;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.ML.Trainers
+{
+    /// <summary>
+    /// Extension methods and utilities for instantiating FFM trainer estimators inside statically typed pipelines.
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineStatic.cs]: >FFM trai [](start = 58, length = 8)

<see cref>

File: src/Microsoft.ML.Data/Evaluators/EvaluatorStaticExtensions.cs [hunk 4]
```diff
@@ -83,5 +83,89 @@ public static class EvaluatorStaticExtensions
...
+            var env = StaticPipeUtils.GetEnvironment(data);
+            Contracts.AssertValue(env);
+            env.CheckValue(label, nameof(label));
+            env.CheckValue(pred, nameof(pred));
+            env.CheckParam(topK >= 0, nameof(topK), "Must not be negative.");
+
+            var indexer = StaticPipeUtils.GetIndexer(data);
+            string labelName = indexer.Get(label(indexer.Indices));
+            (var scoreCol, var predCol) = pred(indexer.Indices);
+            Contracts.CheckParam(scoreCol != null, nameof(pred), "Indexing delegate resulted in null score column.");
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/EvaluatorStaticExtensions.cs]: >Indexing delegate res [](start = 66, length = 21)

is it time to have a Contracts.CheckIndexingDelegate() :)

File: src/Microsoft.ML.Data/Evaluators/EvaluatorStaticExtensions.cs [hunk 5]
```diff
@@ -83,5 +83,89 @@ public static class EvaluatorStaticExtensions
...
+
+        private sealed class TrivialRegressionLossFactory : ISupportRegressionLossFactory
+        {
+            private readonly IRegressionLoss _loss;
+            public TrivialRegressionLossFactory(IRegressionLoss loss) => _loss = loss;
+            public IRegressionLoss CreateComponent(IHostEnvironment env) => _loss;
+        }
+
+        /// <summary>
+        /// Evaluates scored multiclass classification data.
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/EvaluatorStaticExtensions.cs]: >multiclass [](start = 29, length = 10)

regression

File: src/Microsoft.ML.Data/Evaluators/EvaluatorStaticExtensions.cs [hunk 6]
```diff
@@ -83,5 +83,89 @@ public static class EvaluatorStaticExtensions
...
+        /// </summary>
+        /// <typeparam name="T">The shape type for the input data.</typeparam>
+        /// <param name="ctx">The regression context.</param>
+        /// <param name="data">The data to evaluate.</param>
+        /// <param name="label">The index delegate for the label column.</param>
+        /// <param name="score">The index delegate for predicted score column.</param>
+        /// <param name="loss">Potentially custom loss function. If left unspecified defaults to <see cref="SquaredLoss"/>.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Evaluators/EvaluatorStaticExtensions.cs]: >Potentially custom loss function [](start = 31, length = 32)

maybe just simpler: The loss function?

File: src/Microsoft.ML.Data/Training/TrainContext.cs [hunk 7]
```diff
@@ -81,7 +81,7 @@ public sealed class BinaryClassificationContext : TrainContextBase
         /// For trainers for performing binary classification.
         /// </summary>
         /// <remarks>
-        /// Component authors that have written binary classification.
+        /// Component authors that have written binary classification. They are great people.
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Training/TrainContext.cs]: >They are great people. [](start = 71, length = 22)

No one will ever doubt; you can remove the comment :)

File: src/Microsoft.ML.Data/Training/TrainContext.cs [hunk 8]
```diff
@@ -141,4 +141,97 @@ internal BinaryClassificationTrainers(BinaryClassificationContext ctx)
...
         }
     }
+
+    /// <summary>
+    /// The central context for multiclass classification trainers.
+    /// </summary>
+    public sealed class MulticlassClassificationContext : TrainContextBase
+    {
+        /// <summary>
+        /// For trainers for performing multiclass classification.
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Training/TrainContext.cs]: >For trainers for performing multiclass classification. [](start = 12, length = 54)

this is great too.

File: src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineStatic.cs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        /// <param name="numIterations">Number of training iterations.</param>
+        /// <param name="numLatentDimensions">Latent space dimensions.</param>
+        /// <param name="advancedSettings">A delegate to set more settings.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}.Fit(DataView{TTupleInShape})"/> method is called on the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the model that was trained.  Note that this action cannot change the result in any way; it is only a way for the caller to
+        /// be informed about what was learnt.</param>
+        /// <returns>The predicted output.</returns>
+        public static (Scalar<float> score, Scalar<bool> predictedLabel) FieldAwareFactorizationMachine(this BinaryClassificationContext.BinaryClassificationTrainers ctx,
```
Comment (sfilipi) [inline src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineStatic.cs]: >FieldAwareFactorizationMachine [](start = 73, length = 30)

Shall we go with FAFM, if we are doing the acronym for  SDCA?

File: test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj [hunk 10]
```diff
@@ -7,6 +7,7 @@
     <ProjectReference Include="..\..\src\Microsoft.ML.ImageAnalytics\Microsoft.ML.ImageAnalytics.csproj" />
     <ProjectReference Include="..\..\src\Microsoft.ML.StandardLearners\Microsoft.ML.StandardLearners.csproj" />
     <ProjectReference Include="..\Microsoft.ML.TestFramework\Microsoft.ML.TestFramework.csproj" />
+    <NativeAssemblyReference Include="FactorizationMachineNative" />
```
Comment (sfilipi) [inline test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: >NativeAssemblyReference [](start = 5, length = 23)

want to move it close to line 17?

Comment (TomFinley) [inline test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: >FactorizationMachineNative [](start = 38, length = 26)

Hmmm. Opportunistic fix for some other issue?

Comment (Zruty0) [reply test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: No, it's that FFM requires FFMNative, and I added a test for it in static training

---
In reply to: [219902222](https://github.com/dotnet/machinelearning/pull/993#discussion_r219902222) [](ancestors = 219902222)
